infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -106,6 +107,49 @@ TransferSegmentRangeSingle getUnknownAddressesInRangeForRange(TransferSegmentRan
                 .build();
     }
 
+    static boolean sequential(long current, long next) {
+        return current + 1 == next;
+    }
+
+    static Stream<List<Long>> splitNonSequentialAddresses(List<Long> batch) {
+        if (batch.size() < 2) {
+            return Stream.of(batch);
+        }
+        List<List<Long>> lists = new ArrayList<>();
+        List<Long> sequentialList = new ArrayList<>();
+
+        sequentialList.add(batch.get(0));
+        for (int i = 0; i < batch.size() - 1; i++) {
+            long current = batch.get(i);
+            long next = batch.get(i + 1);
+            if (sequential(current, next)) {
+                sequentialList.add(next);
+            } else {
+                lists.add(sequentialList);
+                sequentialList = new ArrayList<>();
+                sequentialList.add(next);
+            }
+        }
+        lists.add(sequentialList);
+        return lists.stream();
+    }
+
+    /**
+     * There are cases in which the unknownAddresses can consist of non-sequential segments.
+     * Make sure that each batch is at least of size batchSize and all addresses in a batch are
+     * sequential.
+     *
+     * @param unknownAddresses List of unknown addresses
+     * @param batchSize        Max size of a batch
+     * @return Batched, sequential addresses
+     */
+    static List<List<Long>> partitionSequentialAddresses(List<Long> unknownAddresses, int batchSize) {
+        return Lists.partition(unknownAddresses, batchSize)
+                .stream()
+                .flatMap(StateTransferManager::splitNonSequentialAddresses)
+                .collect(Collectors.toList());
```

Comment:
I think this method level can be avoided and can be merged with the original code block as mentioned in the overall comment.

Reply:
Fixed